File: JavaOpenDataAccess/src/OpenDataAccess/Entity/EntityType.java
package OpenDataAccess.Entity;

import OpenDataAccess.LambdaInterface.IFunction2;
import OpenDataAccess.Utility.Common;

import java.util.ArrayList;
import java.util.List;

public class EntityType {
    public String Name;
    public String TableName;
    public String WithSql;
    public String PrimaryKey;
    public boolean IsGet;
    public boolean IsPost;
    public boolean IsPut;
    public boolean IsDelete;
    public List<String> NoSelectNameList;
    public LogAttribute LogAttribute;
    public List<Property> Properties;
    public List<IFunction2<IValidate, IEntityData, String>> InsertValidateList;
    public List<IFunction2<IValidate, IEntityData, String>> UpdateValidateList;
    public List<IFunction2<IValidate, IEntityData, String>> DeleteValidateList;

    public EntityType() {
        this.InsertValidateList = new ArrayList<>();
        this.UpdateValidateList = new ArrayList<>();
        this.DeleteValidateList = new ArrayList<>();
    }

    public static List<EntityType> EntityTypeList = new ArrayList<>();

    public static EntityType GetEntityType(String name, boolean blClone) {
        if (Common.IsNullOrEmpty(name)) return null;

        EntityType entityType = Common.GetFirstOrDefault(EntityType.class, EntityTypeList, where -> where.Name.toLowerCase().equals(name.toLowerCase()));
        if (entityType != null && blClone) {
            EntityType newEntityType = new EntityType();
            newEntityType.DeleteValidateList = entityType.DeleteValidateList;
            newEntityType.InsertValidateList = entityType.InsertValidateList;
            newEntityType.Name = entityType.Name;
            newEntityType.PrimaryKey = entityType.PrimaryKey;
            newEntityType.IsDelete = entityType.IsDelete;
            newEntityType.IsGet = entityType.IsGet;
            newEntityType.IsPost = entityType.IsPost;
            newEntityType.IsPut = entityType.IsPut;
            newEntityType.LogAttribute = entityType.LogAttribute;
            if (entityType.Properties != null) {
                List<Property> propertyList = new ArrayList<>();
                entityType.Properties.forEach(a -> {
                    Property p = new Property();
                    p.Name = a.Name;
                    p.ParameterName = a.ParameterName;
                    p.Type = a.Type;
                    p.IsSelect = a.IsSelect;
                    propertyList.add(p);
                });
                newEntityType.Properties = propertyList;
            }
            newEntityType.TableName = entityType.TableName;
            newEntityType.WithSql = entityType.WithSql;
            newEntityType.UpdateValidateList = entityType.UpdateValidateList;
            return newEntityType;
        }
        return entityType;
    }

    public static <T extends IEntity> EntityType GetEntityType(Class<T> cls) throws InstantiationException, IllegalAccessException {
        EntityType entityType = EntityType.GetEntityType(cls.getName(), false);
        if (entityType == null) {
            EntityType.SetEntityType(cls);
            entityType = EntityType.GetEntityType(cls.getName(), false);
        }
        return entityType;
    }

    public static <T extends IEntity> void SetEntityType(Class<T> cls) throws InstantiationException, IllegalAccessException {
        if (!Common.Exists(EntityTypeList, exists -> exists.Name.equals(cls.getName()))) {
            T entity = cls.newInstance();
            EntityTypeList.add(entity.ToEntityType());
        }
    }

    public static void SetEntityTypeByType(Class<?> type) throws InstantiationException, IllegalAccessException {
        if (!Common.Exists(EntityTypeList, exists -> exists.Name.equals(type.getName()))) {
            IEntity entity = (IEntity) type.newInstance();
            if (entity != null) EntityTypeList.add(entity.ToEntityType());
        }
    }

    public static void SetEntityTypeByObject(Object obj) {
        if (!Common.Exists(EntityTypeList, exists -> exists.Name.equals(obj.getClass().getName()))) {
            IEntity entity = (IEntity) obj;
            if (entity != null) EntityTypeList.add(entity.ToEntityType());
        }
    }

    public Property GetProperty(String name) {
        return Common.IsNullOrEmpty(name) ? null : Common.GetFirstOrDefault(Property.class, this.Properties, where -> where.Name.trim().toLowerCase().equals(name.trim().toLowerCase()));
    }

    public List<Property> GetPropertyList(List<String> propertyNameList) {
        List<Property> list = new ArrayList<>();

        this.Properties.forEach(p -> {
            if (Common.Exists(propertyNameList, where -> where.trim().toLowerCase().equals(p.Name.trim().toLowerCase()))) {
                list.add(p);
            }
        });

        return list;
    }
}

File: OpenDataAccess/OpenDataAccess.Entity/EntityType.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace OpenDataAccess.Entity
{
    public class EntityType
    {
        public string Name { get; set; }
        public string TableName { get; set; }
        public string PrimaryKey { get; set; }
        public bool IsGet { get; set; }
        public bool IsPost { get; set; }
        public bool IsPut { get; set; }
        public bool IsDelete { get; set; }
        public LogAttribute LogAttribute { get; set; }
        public List<Property> Properties { get; set; }
        public List<Func<IValidate, IEntityData, string>> InsertValidateList { get; set; }
        public List<Func<IValidate, IEntityData, string>> UpdateValidateList { get; set; }
        public List<Func<IValidate, IEntityData, string>> DeleteValidateList { get; set; }

        public EntityType()
        {
            this.InsertValidateList = new List<Func<IValidate, IEntityData, string>>();
            this.UpdateValidateList = new List<Func<IValidate, IEntityData, string>>();
            this.DeleteValidateList = new List<Func<IValidate, IEntityData, string>>();
        }

        public static List<EntityType> EntityTypeList = new List<EntityType>();

        public static EntityType GetEntityType(string name, bool blClone = false)
        {
            if (string.IsNullOrEmpty(name))
            {
                return null;
            }
            EntityType entityType = EntityTypeList.Where(where => where.Name.Trim().ToLower() == name.Trim().ToLower()).FirstOrDefault();
            if (entityType != null && blClone)
            {
                EntityType newEntityType = new EntityType();
                newEntityType.DeleteValidateList = entityType.DeleteValidateList;
                newEntityType.InsertValidateList = entityType.InsertValidateList;
                newEntityType.Name = entityType.Name;
                newEntityType.PrimaryKey = entityType.PrimaryKey;
                newEntityType.IsDelete = entityType.IsDelete;
                newEntityType.IsGet = entityType.IsGet;
                newEntityType.IsPost = entityType.IsPost;
                newEntityType.IsPut = entityType.IsPut;
                newEntityType.LogAttribute = entityType.LogAttribute;
                if (entityType.Properties != null)
                {
                    newEntityType.Properties = (from a in entityType.Properties
                                                select new Property()
                                                {
                                                    Name = a.Name,
                                                    ParameterName = a.ParameterName,
                                                    Type = a.Type
                                                }).ToList();
                }
                newEntityType.TableName = entityType.TableName;
                newEntityType.UpdateValidateList = entityType.UpdateValidateList;
                return newEntityType;
            }
            return entityType;
        }

        public static EntityType GetEntityType<T>() where T : IEntity
        {
            EntityType entityType = EntityType.GetEntityType(typeof(T).Name);
            if (entityType == null)
            {
                EntityType.SetEntityType<T>();
                entityType = EntityType.GetEntityType(typeof(T).Name);
            }
            return entityType;
        }

        public static void SetEntityType<T>() where T : IEntity
        {
            if (!EntityTypeList.Exists(exists => exists.Name == typeof(T).Name))
            {
                T entity = Activator.CreateInstance<T>();
                EntityTypeList.Add(entity.ToEntityType());
            }
        }

        public static void SetEntityType(Type type)
        {
            if (!EntityTypeList.Exists(exists => exists.Name == type.Name))
            {
                IEntity entity = Activator.CreateInstance(type) as IEntity;
                if (entity != null)
                {
                    EntityTypeList.Add(entity.ToEntityType());
                }
            }
        }

        public static void SetEntityType(object obj)
        {
            if (!EntityTypeList.Exists(exists => exists.Name == obj.GetType().Name))
            {
                IEntity entity = obj as IEntity;
                if (entity != null)
                {
                    EntityTypeList.Add(entity.ToEntityType());
                }
            }
        }

        public Property GetProperty(string name)
        {
            return string.IsNullOrEmpty(name) ? null : this.Properties.Where(where => where.Name.Trim().ToLower() == name.Trim().ToLower()).FirstOrDefault();
        }

        public List<Property> GetPropertyList(List<string> propertyNameList)
        {
            return (from property in this.Properties
                    from proeprtyName in propertyNameList
                    where property.Name.Trim().ToLower() == proeprtyName.Trim().ToLower()
                    select property).ToList();
        }

    }
}
